src/API/Management/Rules.php

```diff
@@ -2,107 +2,179 @@
     }
 
     /**
+     * Delete a rule by ID.
+     * Required scope: "delete:rules"
+     *
+     * @param string $id Rule ID to delete.
      *
-     * @param  string $id
      * @return mixed
+     *
+     * @throws CoreException Thrown when $id is empty or not a string.
+     * @throws \Exception Thrown by the HTTP client when there is a problem with the API call.
+     *
+     * @link https://auth0.com/docs/api/management/v2#!/Rules/delete_rules_by_id
      */
     public function delete($id)
     {
-        return $this->apiClient->delete()
-            ->rules($id)
+        if (empty($id) || ! is_string($id)) {
+            throw new CoreException('Empty or invalid "id" parameter.');
+        }
+
+        return $this->apiClient->method('delete')
+            ->addPath('rules', $id)
             ->call();
     }
 
     /**
+     * Create a new Rule.
+     * Required scope: "create:rules"
+     *
+     * @param array $data Dictionary array of keys and values to create a Rule.
      *
-     * @param  array $data
      * @return mixed
+     *
+     * @throws CoreException Thrown when required "script" or "name" fields are missing or empty.
+     * @throws \Exception Thrown by the HTTP client when there is a problem with the API call.
+     *
+     * @link https://auth0.com/docs/api/management/v2#!/Rules/post_rules
+     * @link https://auth0.com/docs/rules/current#create-rules-with-the-management-api
      */
-    public function create($data)
+    public function create(array $data)
     {
-        return $this->apiClient->post()
-            ->rules()
-            ->withHeader(new ContentType('application/json'))
+        if (empty($data['name'])) {
+            throw new CoreException('Missing required "name" field.');
+        }
+
+        if (empty($data['script'])) {
+            throw new CoreException('Missing required "script" field.');
+        }
+
+        return $this->apiClient->method('post')
+            ->addPath('rules')
             ->withBody(json_encode($data))
             ->call();
     }
 
     /**
+     * Update a Rule by ID.
+     * Required scope: "update:rules"
+     *
+     * @param string $id   Rule ID to delete.
+     * @param array  $data Rule data to update.
      *
-     * @param  string $id
-     * @param  array  $data
      * @return mixed
+     *
+     * @throws CoreException Thrown when $id is empty or not a string or if $data is empty.
+     * @throws \Exception Thrown by the HTTP client when there is a problem with the API call.
      */
-    public function update($id, $data)
+    public function update($id, array $data)
     {
-        return $this->apiClient->patch()
-            ->rules($id)
-            ->withHeader(new ContentType('application/json'))
+        if (empty($id) || ! is_string($id)) {
+            throw new CoreException('Empty or invalid "id" parameter.');
+        }
+
+        if (empty($data)) {
+            throw new CoreException('Empty "data" parameter; nothing to do.');
```

Comment:
Can you make your errors consistent? In the `create` method you have 
`Missing required "script" field.` 
in this one it's 
`Empty "data" parameter; nothing to do.`

If it's empty or invalid it's still invalid, as these are essential I would just have something simple like
`{param} is missing`

Reply:
It's indicating 2 different things. In `create`, we require 2 specific fields in that `$data` array to create a Rule. In `update`, we don't require any specific fields but if it's empty then there is no reason to send the request. The error messages reflect that.

Reply:
Jumping in here. 

1. `Empty or invalid "id" parameter.` I agree with martin. That's "invalid" anyway. You can have a helper function somewhere that receives the value and the name of the parameter [like I do](https://github.com/auth0/auth0-java/blob/master/src/main/java/com/auth0/utils/Asserts.java#L8) and so keep a universal message like "Invalid {param name} parameter" somewhere.

2. For the PATCH I know it makes no sense at all to make the call if the value is empty, but it's not that the call will fail with an error from the server. AFAIK It just won't do anything. Anyway, I also agree messages should be similar. If you are going to consider this an error scenario then just use the same message as in (1).

Reply:
Thanks for the discussion here but I still don't see this as being inconsistent. I'm fine with changing `Empty or invalid` to just 'Invalid` but where the message is different, the condition is different as well. 

- For `get`, `delete`, and `create`, the ID parameter needs to have a value (not `null` or `""`) or it's empty. It needs to also be a string (not `[12345678]` or `SomeClassName`) or it's invalid. Again, I'm fine with changing the exact wording.
- For `create`, that forces an array to be passed so we know the type is right. But there are 2 specific keys/fields that it needs. If I say `Invalid $data parameter`, then I'm leaving out helpful information for debugging. I could say `Invalid $data parameter ("name" and "script" keys are required)` ... does that seem more consistent to you guys?